core-spi-test/src/main/java/org/ehcache/internal/store/StoreEventListenerTest.java

```diff
@@ -0,0 +1,333 @@
+    TestTimeSource timeSource = new TestTimeSource();
+    Store<K, V> kvStore = factory.newStore(factory.newConfiguration(factory.getKeyType(), factory.getValueType(),
+        null, null, null, Expirations.timeToLiveExpiration(new Duration(1, TimeUnit.MILLISECONDS))), timeSource);
+    TestStoreEventListener listener = new TestStoreEventListener();
+    kvStore.enableStoreEventNotifications(listener);
+    kvStore.put(k, v);
+    timeSource.advanceTime(1);
+    kvStore.compute(k, new BiFunction<K, V, V>() {
+      @Override
+      public V apply(K mappedKey, V mappedValue) {
+        return v2;
+      }
+    });
+    assertThat(kvStore.get(k).value(), is(v2));
+    assertThat(listener.expired, hasItem(k));
+  }
+
+  @SuppressWarnings({ "rawtypes", "unchecked" })
+  @SPITest
+  public void testComputeIfAbsentOnExpiration() throws Exception {
+    TestTimeSource timeSource = new TestTimeSource();
+    Store<K, V> kvStore = factory.newStore(factory.newConfiguration(factory.getKeyType(), factory.getValueType(),
+        null, null, null, Expirations.timeToLiveExpiration(new Duration(1, TimeUnit.MILLISECONDS))), timeSource);
+    TestStoreEventListener listener = new TestStoreEventListener();
+    kvStore.enableStoreEventNotifications(listener);
+    kvStore.put(k, v);
+    timeSource.advanceTime(1);
+
+    kvStore.computeIfAbsent(k, new Function<K, V>() {
+      @Override
+      public V apply(K mappedKey) {
+        return v2;
+      }
+    });
+    assertThat(kvStore.get(k).value(), is(v2));
+    assertThat(listener.expired, hasItem(k));
+  }
+
+  @SuppressWarnings({ "rawtypes", "unchecked" })
+  @SPITest
+  public void testComputeIfPresentOnExpiration() throws Exception {
+    TestTimeSource timeSource = new TestTimeSource();
+    Store<K, V> kvStore = factory.newStore(factory.newConfiguration(factory.getKeyType(), factory.getValueType(),
+        null, null, null, Expirations.timeToLiveExpiration(new Duration(1, TimeUnit.MILLISECONDS))), timeSource);
+    TestStoreEventListener listener = new TestStoreEventListener();
+    kvStore.enableStoreEventNotifications(listener);
+    kvStore.put(k, v);
+    timeSource.advanceTime(1);
+
+    kvStore.computeIfPresent(k, new BiFunction<K, V, V>() {
+      @Override
+      public V apply(K mappedKey, V mappedValue) {
+        throw new AssertionError();
+      }
+    });
+    assertThat(kvStore.get(k), is(nullValue()));
+    assertThat(listener.evicted, hasItem(k));
+  }
+
+  @SuppressWarnings({ "rawtypes", "unchecked" })
+  @SPITest
+  public void testPutOnEviction() throws Exception {
+    Store<K, V> kvStore = factory.newStore(factory.newConfiguration(factory.getKeyType(), factory.getValueType(),
+        1L, null, null));
+    TestStoreEventListener listener = new TestStoreEventListener();
+    kvStore.enableStoreEventNotifications(listener);
+    kvStore.put(k, v);
+    kvStore.put(k2, v2);
+    Map<K, V> map = convertToMap(kvStore.iterator());
+    assertThat(map.size(), is(1));
```

Comment:
This applies to probably all of the eviction tests (if we think this is the right thing to do).